src/virtio/virtio_mmio.c

```diff
@@ -0,0 +1,327 @@
+
+    switch (offset) {
+        case REG_RANGE(REG_VIRTIO_MMIO_MAGIC_VALUE, REG_VIRTIO_MMIO_VERSION):
+            reg = VIRTIO_MMIO_DEV_MAGIC;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_VERSION, REG_VIRTIO_MMIO_DEVICE_ID):
+            reg = VIRTIO_MMIO_DEV_VERSION;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_DEVICE_ID, REG_VIRTIO_MMIO_VENDOR_ID):
+            reg = emul_handler->data.DeviceID;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_VENDOR_ID, REG_VIRTIO_MMIO_DEVICE_FEATURES):
+            reg = emul_handler->data.VendorID;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_DEVICE_FEATURES, REG_VIRTIO_MMIO_DEVICE_FEATURES_SEL):
+            success = emul_handler->funs->get_device_features(emul_handler, &reg);
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_QUEUE_NUM_MAX, REG_VIRTIO_MMIO_QUEUE_NUM):
+            reg = QUEUE_SIZE;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_QUEUE_READY, REG_VIRTIO_MMIO_QUEUE_NOTIFY):
+            if (emul_handler->data.QueueSel < VIRTIO_MMIO_NET_NUM_VIRTQUEUE) {
+                reg = emul_handler->vqs[emul_handler->data.QueueSel].ready;
+            }
+            // simply ignore wrong queue number
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_INTERRUPT_STATUS, REG_VIRTIO_MMIO_INTERRUPT_ACK):
+            reg = emul_handler->data.InterruptStatus;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_STATUS, REG_VIRTIO_MMIO_QUEUE_DESC_LOW):
+            success = handle_virtio_mmio_get_status_flag(emul_handler, &reg);
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_CONFIG_GENERATION, REG_VIRTIO_MMIO_CONFIG):
+            /* ConfigGeneration will need to be update every time when the backend changes any of
+             * the device config. Currently we only have virtio net backend that doesn't do any update
+             * on the device config, so the ConfigGeneration is always 0. I left this comment
+             * here as a reminder. */
+            // reg = emul_handler->data.ConfigGeneration;
+            break;
+
+        case REG_RANGE(REG_VIRTIO_MMIO_CONFIG, REG_VIRTIO_MMIO_CONFIG + 0x100):
+            success = emul_handler->funs->get_device_config(emul_handler, offset, &reg);
+
+            // uint32_t mask = fault_get_data_mask(fault_addr, fsr);
+            // printf("\"%s\"|VIRTIO MMIO|INFO: device config offset 0x%x, value 0x%x, mask 0x%x\n", sel4cp_name, offset, reg & mask, mask);
+            break;
+
+        default:
+            printf("VIRTIO MMIO|INFO: unknown register 0x%x.\n", offset);
+            success = 0;
+    }
+
+    // we are expected to reply even on error
+    uint32_t mask = fault_get_data_mask(fault_addr, fsr);
+    int ret = fault_advance(regs, fault_addr, fsr, reg & mask);
+    assert(ret);
+
+    return success;
+}
+
+static bool handle_virtio_mmio_reg_write(uint64_t fault_addr, uint64_t fsr, seL4_UserContext *regs)
+{
+
```

Comment:
It would be cool if we had debug-logging (similar to the VGIC logging that is off by default) for every virtIO access, so e.g for reading we can print out the address/register, for write print out value and address/register